source/common/config/grpc_stream.h

```diff
@@ -109,11 +96,33 @@ class GrpcStream : public Grpc::TypedAsyncStreamCallbacks<ResponseProto>,
 private:
   void drainRequests() {
     ENVOY_LOG(trace, "draining discovery requests {}", request_queue_.size());
-    while (!request_queue_.empty() && checkRateLimitAllowsDrain(request_queue_.size())) {
+    while (!request_queue_.empty() && checkRateLimitAllowsDrain()) {
       // Process the request, if rate limiting is not enabled at all or if it is under rate limit.
       sendDiscoveryRequest(request_queue_.front());
       request_queue_.pop();
     }
+    // Although request_queue_.push() happens elsewhere, the only time the queue is non-transiently
+    // non-empty is when it remains non-empty after a drain attempt. (The push() doesn't matter
+    // because we always attempt this drain immediately after the push). Basically, a change in
+    // queue length is not "meaningful" until it has persisted until here. Since we are constantly
+    // doing set(0), we need the if(>0 || used) to keep this stat from being wrongly marked
```

Comment:
What do you mean by "constantly doing set(0)"? Where does that happen?

Reply:
Every DiscoveryRequest goes through this queue+drain process, even if it's not actually going to be queued. I did write it kind of weird, didn't I: I really meant "if not for this conditional we would be doing it constantly." Clarified the comment.